addons/purchase_requisition/models/purchase_requisition.py

```diff
@@ -98,11 +60,58 @@ def _onchange_vendor(self):
             }
             return {'warning': warning}
 
+    @api.depends('vendor_id')
+    def _compute_currency_id(self):
+        for requisition in self:
+            if not requisition.vendor_id or not requisition.vendor_id.property_purchase_currency_id:
+                requisition.currency_id = requisition.company_id.currency_id.id
+            else:
+                requisition.currency_id = requisition.vendor_id.property_purchase_currency_id.id
+
     @api.depends('purchase_ids')
     def _compute_orders_number(self):
         for requisition in self:
             requisition.order_count = len(requisition.purchase_ids)
 
+    @api.constrains('date_start', 'date_end')
+    def _check_dates(self):
+        invalid_requsitions = self.filtered(lambda r: r.date_end and r.date_start and r.date_end < r.date_start)
+        if invalid_requsitions:
+            raise ValidationError(_(
+                "End date cannot be earlier than start date. Please check dates for agreements: %s", ', '.join(invalid_requsitions.mapped('name'))
+            ))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        defaults = self.default_get(['requisition_type', 'company_id'])
+        for vals in vals_list:
+            requisition_type = vals.get('requisition_type', defaults['requisition_type'])
+            company_id = vals.get('company_id', defaults['company_id'])
+            if requisition_type == 'blanket_order':
+                vals['name'] = self.env['ir.sequence'].with_company(company_id).next_by_code('purchase.requisition.blanket.order')
+            else:
+                vals['name'] = self.env['ir.sequence'].with_company(company_id).next_by_code('purchase.requisition.purchase.template')
+        return super().create(vals_list)
+
+    def write(self, vals):
+        requisition_to_rename = self.env['purchase.requisition']
+        if 'requisition_type' in vals or 'company_id' in vals:
```

Comment:
The field is readonly in the view for running requisition, but I would add a check here on the state to be sure

Reply:
Good point, I added into a Usererror if a requisition to rename is not in state draft after the write